### PR TITLE
dex

### PR DESCRIPTION
**Pull request by mhrabovcin**

### CODE REVIEW

File: templates/dex-k8s-authenticator.yaml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+kind: Addon
+metadata:
+  name: dex-k8s-authenticator
+  namespace: kubeaddons
+spec:
+  kubernetes:
+    minSupportedVersion: v1.14.0
+  chartReference:
+    chart: raw.githubusercontent.com/dex-k8s-authenticator
+    # TODO(mh): If this is going to be used figure out the production URL
```
Comment (shaneutt) [inline templates/dex-k8s-authenticator.yaml]: If its agreeable for you @mhrabovcin, I would like to have a JIRA created for this TODO so that we can better keep track of this need (currently, we're not allowed to utilize github issues for kubeaddons work organization). Does that sound :heavy_check_mark: ?

File: templates/dex.yaml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+spec:
+  kubernetes:
+    minSupportedVersion: v1.14.0
+  chartReference:
+    chart: stable/dex
+    version: 1.4.0
+    values: |
+      ---
+      image: mhrabovcin/dex
+      imageTag: "d2f240c9c72682913eaff4fd85d119d7d2d0f240"
```
Comment (shaneutt) [inline templates/dex.yaml]: Might be a nice cordiality to add a comment above this with the entire URL pointing to this revision, just makes it easy for someone to click while viewing the code. What do you think?

File: templates/dex.yaml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+          tlsCert: /etc/dex/tls/https/server/tls.crt
+          tlsKey: /etc/dex/tls/https/server/tls.key
+        grpc:
+          addr: 0.0.0.0:5000
+          tlsCert: /etc/dex/tls/grpc/server/tls.crt
+          tlsKey: /etc/dex/tls/grpc/server/tls.key
+          tlsClientCA: /etc/dex/tls/grpc/ca/tls.crt
+        oauth2:
+          skipApprovalScreen: true
+
```
Comment (shaneutt) [inline templates/dex.yaml]: nit: whitespace
```suggestion
```

File: templates/dex.yaml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        staticClients:
+        # This `id` must by in sync with `dex-k8s-authenticator.yaml` value as well as
```
Comment (shaneutt) [inline templates/dex.yaml]: comments++ :+1:

File: templates/traefik.yaml [hunk 5]
```diff
@@ -36,6 +36,10 @@ spec:
         enabled: true
         enforced: true
         generateTLS: true
+        # dex service is exposed with TLS certificate signed by self signed root
+        # Dex CA certificate.
+        # TODO(mh): Figure out if we can configure CA certs per endpoint/ingress
```
Comment (shaneutt) [inline templates/traefik.yaml]: Is this another place it would be good to have a JIRA created for follow up (and linked here in the comments)?

File: templates/dex.yaml [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  namespace: kubeaddons
+spec:
+  kubernetes:
+    minSupportedVersion: v1.14.0
+  chartReference:
+    chart: stable/dex
+    version: 1.4.0
+    values: |
+      ---
+      image: mhrabovcin/dex
```
Comment (hectorj2f) [inline templates/dex.yaml]: shouldn't we move this docker image under mesosphere ?

File: templates/dex.yaml (hunk 2 above)
Comment (hectorj2f) [reply templates/dex.yaml]: can we get a version tag even if it is temporal ?

File: templates/dex.yaml [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+          tlsClientCA: /etc/dex/tls/grpc/ca/tls.crt
+        oauth2:
+          skipApprovalScreen: true
+
+        staticClients:
+        # This `id` must by in sync with `dex-k8s-authenticator.yaml` value as well as
+        # kube-apiserver flag `oidc-client-id`.
+        - id: kube-apiserver
+          redirectURIs:
+            - 'https://dex-k8s-authenticator-kubeaddons.kubeaddons.svc.cluster.local:5555/dex-k8s-authenticator/callback/kubernetes-cluster'
```
Comment (hectorj2f) [inline templates/dex.yaml]: Note that `cluster.local` could vary although we are not setting it at this moment.

File: templates/dex-k8s-authenticator.yaml [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        #tlsKey: /path/to/dex-client.key
+        clusters:
+        - name: kubernetes-cluster
+          short_description: "Kubernetes cluster"
+          description: "Kubernetes cluster authenticator"
+          # client_secret: value is generated automatically - see `setupDexK8sAuthenticator()`
+          # This `client_id` must by in sync with `dex.yaml` value as well as
+          # kube-apiserver flag `oidc-client-id`.
+          client_id: kube-apiserver
+          issuer: https://dex-kubeaddons.kubeaddons.svc.cluster.local:8080/dex
```
Comment (hectorj2f) [inline templates/dex-k8s-authenticator.yaml]: This is an example where this settings cannot work without the dex settings defined below.

File: templates/dex.yaml (hunk 2 above)
Comment (shaneutt) [reply templates/dex.yaml]: Indeed, it would be nice to just tag anything we're going to use for this purpose for the time being.

File: templates/dex.yaml (hunk 7 above)
Comment (mhrabovcin) [reply templates/dex.yaml]: I'll add the following comment:

```
            # This URI is just a placeholder and it will be replaced during the
            # deployment with URL pointing to the traefik ingress public
            # load balancer.
```

File: templates/dex-k8s-authenticator.yaml (hunk 8 above)
Comment (mhrabovcin) [reply templates/dex-k8s-authenticator.yaml]: What would be a preferred way to enforce dependency between `dex` and `dex-k8s-authenticator`?

File: templates/dex.yaml (hunk 6 above)
Comment (mhrabovcin) [reply templates/dex.yaml]: I am going to propose changes to upstream `dex` repository and ideally, we'll be running off official container. If we won't get changes upstream in the time I am planning to push custom version to our `mesosphere` namespace.

File: templates/traefik.yaml (hunk 5 above)
Comment (mhrabovcin) [reply templates/traefik.yaml]: I've created https://jira.mesosphere.com/browse/DCOS-56033

File: templates/dex.yaml (hunk 6 above)
Comment (shaneutt) [reply templates/dex.yaml]: As this is high prio and it appears a suitable answer was provided for the concern, I'm going to mark this as resolved but please reopen and flag request changes if this is not the case.

File: templates/dex-k8s-authenticator.yaml (hunk 8 above)
Comment (shaneutt) [reply templates/dex-k8s-authenticator.yaml]: I believe for the moment and in the interest of time we should skip on worrying about this dependency management issue and allow a manual approach in the interim. I have created [https://jira.mesosphere.com/browse/DCOS-56045](DCOS-56045) to cover the next iteration on this wherein we need to automate this.

File: templates/dex-k8s-authenticator.yaml [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  namespace: kubeaddons
+spec:
+  kubernetes:
+    minSupportedVersion: v1.14.0
+  chartReference:
+    chart: raw.githubusercontent.com/dex-k8s-authenticator
+    # TODO(mh): If this is going to be used figure out the production URL
+    # See: https://github.com/mintel/dex-k8s-authenticator/issues/70
+    repo: https://raw.githubusercontent.com/mintel/dex-k8s-authenticator/feat-helm-repo/charts/repo/stable
+    version: 1.0.0
```
Comment (hectorj2f) [inline templates/dex-k8s-authenticator.yaml]: Shouldn't the version be `version: 1.1.1` as in mesosphere/charts ?

Comment (hectorj2f) [reply templates/dex-k8s-authenticator.yaml]: yes, it has to be that version https://github.com/mesosphere/charts/blob/master/docs/staging/index.yaml#L21

Comment (mhrabovcin) [reply templates/dex-k8s-authenticator.yaml]: Yes, I am still working on an update to this PR where I'll point dex-k8s-authenticator to our custom chart repo and as well move use dex container from `mesosphere/dex` repo.

Comment (hectorj2f) [reply templates/dex-k8s-authenticator.yaml]: 👍

File: templates/traefik.yaml [hunk 10]
```diff
@@ -39,6 +39,12 @@ spec:
         enabled: true
         enforced: true
         generateTLS: true
+        # dex service is exposed with TLS certificate signed by self signed root
+        # Dex CA certificate. It is not clear if traefik supports configuring
+        # trusted certificates per backend. This should be investiaged in a
+        # separate issue.
+        # See: https://jira.mesosphere.com/browse/DCOS-56033
+        insecureSkipVerify: true
```
Comment (hectorj2f) [inline templates/traefik.yaml]: the default is `insecureSkipVerify: false` that will affect other usages of traefik :/. 
I don't like that but let's move forward and address it in a different PR. Perhaps we have to set this value for traefik when Dex is enabled.
